core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -516,61 +517,131 @@ class KafkaApis(val requestChannel: RequestChannel,
-        case nle: NotLeaderForPartitionException =>
-          debug("Offset request with correlation id %d from client %s on partition %s failed due to %s".format(
-               correlationId, clientId, topicPartition,nle.getMessage))
-          (topicPartition, new ListOffsetResponse.PartitionData(Errors.forException(nle).code, List[JLong]().asJava))
-        case e: Throwable =>
-          error("Error while responding to offset request", e)
-          (topicPartition, new ListOffsetResponse.PartitionData(Errors.forException(e).code, List[JLong]().asJava))
-      }
-    })
 
-    val mergedResponseMap = responseMap ++ unauthorizedResponseStatus
+        val unauthorizedResponseStatus = unauthorizedRequestInfo.mapValues(_ => {
+          new PartitionData(Errors.TOPIC_AUTHORIZATION_FAILED.code,
+            ListOffsetResponse.UNKNOWN_TIMESTAMP,
+            ListOffsetResponse.UNKNOWN_OFFSET)
+        })
+
+        val responseMap = authorizedRequestInfo.map(elem => {
+          val (topicPartition, timestamp) = elem
+          if (offsetRequest.duplicatePartitions().contains(topicPartition)) {
+            debug(s"OffsetRequest with correlation id $correlationId from client $clientId on partition $topicPartition " +
+                s"failed because the partition is duplicated in the request.")
+            (topicPartition, new PartitionData(Errors.INVALID_REQUEST.code(),
+                                               ListOffsetResponse.UNKNOWN_TIMESTAMP,
+                                               ListOffsetResponse.UNKNOWN_OFFSET))
+          } else {
+            try {
+              // ensure leader exists
+              val localReplica = if (offsetRequest.replicaId != ListOffsetRequest.DEBUGGING_REPLICA_ID)
+                replicaManager.getLeaderReplicaIfLocal(topicPartition.topic, topicPartition.partition)
+              else
+                replicaManager.getReplicaOrException(topicPartition.topic, topicPartition.partition)
+              val found = {
+                fetchOffsetForTimestamp(replicaManager.logManager, topicPartition, timestamp) match {
+                  case Some(timestampOffset) if offsetRequest.replicaId != ListOffsetRequest.CONSUMER_REPLICA_ID ||
+                      timestampOffset.offset <= localReplica.highWatermark.messageOffset =>
+                    timestampOffset
+                  case None =>
+                    TimestampOffset(ListOffsetResponse.UNKNOWN_TIMESTAMP, ListOffsetResponse.UNKNOWN_OFFSET)
+                }
+              }
+              (topicPartition, new ListOffsetResponse.PartitionData(Errors.NONE.code, found.timestamp, found.offset))
+            } catch {
+              // NOTE: UnknownTopicOrPartitionException and NotLeaderForPartitionException are special cased since these error messages
+              // are typically transient and there is no value in logging the entire stack trace for the same
+              case utpe: UnknownTopicOrPartitionException =>
```

Comment:
Some of these seem to be doing the same thing. Can we merge them into a single clause?